ofrak_components/ofrak_components/zip.py

```diff
@@ -61,17 +56,19 @@ class ZipPacker(Packer[None]):
 
     targets = (ZipArchive,)
 
+    ZIP_TOOL = ComponentExternalTool("zip")
```

Comment:
What if, instead, components used tools directly. For example:

```
PIGZ_TOOL = ComponentExternalTool("pigz")
```
And then the unpacker/packer both use the tool directly? This would be cleaner for situations where multiple components use the same tool (This is the case with binary ninja components, and even the gzip unpacker/packers).

Reply:
This makes it impractical to tie dependencies to the components that need them, but that may not be important.

Reply:
Hm, my thinking here was that `get_external_dependencies` could handle that. So, for example of [APK](https://github.com/redballoonsecurity/ofrak/blob/master/ofrak_components/ofrak_components/apk.py), something like

```
UBER_APK_SIGNER = CommandExternalTool(["java", "-jar", "/usr/local/bin/uber-apk-signer.jar"])
APK_TOOL = CommandExternalTool("apktool")

class ApkPacker(Packer):
...
    @classmethod
    def get_expernal_dependencies(cls):
        return UBER_APK_SIGNER, APK_TOOL
```

Reply:
One general question that I have is: why is the "component external tool" management implemented outside of the `Unpacker` and `Packer` classes as opposed to extending them with methods to manage the dependencies? I would expect it's easier to reason about dependencies if they're installed near where they're used. Likewise it might lower the barrier to writing new components.

What are the design justifications for having this additional, separate class?

Reply:
@rbs-jacob 
It makes sense to make a separate class because of the [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle). 

Also, as I point out above, some external dependencies are used across multiple components, and components, such as the `ApkPacker`, can also have multiple external dependencies.

Reply:
Implemented pretty much as suggested here, but using a "pseudo-property" like `targets` for components. `external_dependencies` should be a tuple of external tools. The `AbstractComponent` defaults to providing an empty Tuple for this, I think that makes sense: the pro is that all the components without external dependencies don't need to be marked as such, the downside is that it might be easier for a contributor to miss marking external dependencies on a component because they forgot.